mirrord/protocol/src/codec.rs

```diff
@@ -142,77 +143,37 @@ pub enum DaemonMessage {
     SwitchProtocolVersionResponse(#[bincode(with_serde)] semver::Version),
 }
 
-pub struct ClientCodec {
+pub struct ProtocolCodec<I, O> {
     config: bincode::config::Configuration,
+    /// Phantom just to associate the message types with the struct.
+    /// Implementing a trait to associate the types is not good enough, because we can't blanket
+    /// implement the foreign `Encoder`/`Decoder` traits, so we have to implement them for a
+    /// struct.
+    _phantom_incoming_message: PhantomData<I>,
+    _phantom_outgoing_message: PhantomData<O>,
 }
 
-impl ClientCodec {
-    pub fn new() -> Self {
-        ClientCodec {
-            config: bincode::config::standard(),
-        }
-    }
-}
-
-impl Default for ClientCodec {
-    fn default() -> Self {
-        ClientCodec::new()
-    }
-}
-
-impl Decoder for ClientCodec {
-    type Item = DaemonMessage;
-    type Error = io::Error;
-
-    fn decode(&mut self, src: &mut BytesMut) -> io::Result<Option<Self::Item>> {
-        match bincode::decode_from_slice(&src[..], self.config) {
-            Ok((message, read)) => {
-                src.advance(read);
-                Ok(Some(message))
-            }
-            Err(DecodeError::UnexpectedEnd { .. }) => Ok(None),
-            Err(err) => Err(io::Error::new(io::ErrorKind::Other, err.to_string())),
-        }
-    }
-}
-
-impl Encoder<ClientMessage> for ClientCodec {
-    type Error = io::Error;
+pub type ClientCodec = ProtocolCodec<DaemonMessage, ClientMessage>;
+pub type DaemonCodec = ProtocolCodec<ClientMessage, DaemonMessage>;
 
-    fn encode(&mut self, msg: ClientMessage, dst: &mut BytesMut) -> Result<(), Self::Error> {
-        let encoded = match bincode::encode_to_vec(msg, self.config) {
-            Ok(encoded) => encoded,
-            Err(err) => {
-                return Err(io::Error::new(io::ErrorKind::Other, err.to_string()));
-            }
-        };
-        dst.reserve(encoded.len());
-        dst.put(&encoded[..]);
-
-        Ok(())
-    }
-}
-
-pub struct DaemonCodec {
-    config: bincode::config::Configuration,
-}
-
-impl DaemonCodec {
+impl<I, O> ProtocolCodec<I, O> {
     pub fn new() -> Self {
```

Comment:
```suggestion
    fn new() -> Self {
```

Drop the `pub`, we're supposed to build this with `default`, right?

Reply:
There are many existing places where we build Client/Daemon codecs with `new`, but I now changed them all to default and remove this method altogether, see last commit.